artefactual-sdps/enduro
# Use OneOf in package events

This is a long-standing pull request where I attempted to make use of OneOf in Goa but I've been running into various blockers that were reported to Goa and releases were made as a result. I still think that there may be another bug but it's not a blocker, read more below.

##### How to test

1. Confirm that the UI continuous receiving updates asynchronously, i.e. you don't need to refresh the page to receive data changes.
2. Open the Network tab in your browser and confirm that the WS channel is open and you see messages being received.

##### Description

This pull request introduces the use of [OneOf](https://pkg.go.dev/goa.design/goa/v3@v3.14.0/dsl#OneOf) for a more efficient way of expressing the monitor event data. In our current design of the type [type](https://github.com/artefactual-sdps/enduro/blob/e68a7c750e27f27a6a1fc5e0056230c626a3635a/internal/api/design/package_monitor.go#L7-L56) we simply list a set of nullable attributes. In other words, we have a message with many fields where at most one field will be set at the same time. Enforcing this behavior with OneOf is a good practice.

When connected to the WebSocket, messages look as follows:

```json
{"event":{"Type":"monitor_ping_event","Value":"{\"Message\":\"Ping\"}"}}
```

It is more succint if you compare it with the previous message:

```json
{"MonitorPingEvent":{"Message":"Ping"},"PackageCreatedEvent":null,"PackageUpdatedEvent":null,"PackageStatusUpdatedEvent":null,"PackageLocationUpdatedEvent":null,"PreservationActionCreatedEvent":null,"PreservationActionUpdatedEvent":null,"PreservationTaskCreatedEvent":null,"PreservationTaskUpdatedEvent":null}
```

What's interesting about the new structure generated by Goa:
1. The event is JSON-encoded but also the inner value is encoded again,
2. Instead of using snake_case like in other areas of the API, this uses SnakeCase, e.g. I would expect "Message" to read as "message". I'm solving that problem on the client for now as follows:

```ts
handleEvent(event: api.MonitorEventEvent) {
  const json = JSON.parse(event.value);
  const value = mapKeys(json, (value, key) => snakeCase(key));
  handlers[event.type](value);
}
```

mapKeys is used to convert attributes from snakeCase to snake_case. I suspect this is a bug in Goa, but I have started a conversation with the maintainers to confirm whether that's the case or I'm just doing something wrong. Generally, Goa encodes JSON using snake_case for attributes, so this is definitely an anomaly. It's easy for me to do the conversion from the client, but the generated spec reads that should be snake_case. This is good enough until we figure out if we can address it on the backend.
